File: Leetcode_JS/good-nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function(root) {
    // node x cannot be less than the root

    if (!root) return 0
    let count = 0
    let stack = []
    stack.push([root, -Infinity])

    while (stack.length) {
        let [node,max] = stack.pop()
        const {left, right, val} = node
        console.log(node, left)
        if (max <= val) {
            count++
            max = val
        }

        if (left) stack.push([left,max])
        if (right) stack.push([right,max])
    }
    return count
};

File: Leetcode_Python/good-nodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def goodNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        def dfs(node, curr_max):
            if not node:
                return
            if node.val >= curr_max:
                count[0]+=1
                curr_max = node.val
            dfs(node.left, curr_max)
            dfs(node.right, curr_max)

        count= [0]
        dfs(root, root.val)
            
        return count[0]